From 7650b8f69f78a830b1580aa775e91fa277c0e132 Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Tue, 30 May 2023 19:09:20 +0800
Subject: [PATCH 32/53] [Huawei] mm/sharepool: Adjust delete and create process
 for sp_area

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

1. Extract sp_group_delete_area() that delete a sp_area from sp_group.
   We should promise that the sp_area has been deleted from the sp_group
   before we free the sp_area.
2. Update spa_stat when we add/delete a sp_area to/from a sp_group.
   Because we need to ensure the existence of the sp_group and the spg
   could be deleted before. (This would not happen now because of the
   implementation, but this is not guaranteed logically)

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 42 ++++++++++++++++++++++++++++--------------
 1 file changed, 28 insertions(+), 14 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index 8789e1da6769..81d71c051bf3 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -763,7 +763,6 @@ static void spa_inc_usage(struct sp_area *spa)
 		atomic64_add(ALIGN(size, PMD_SIZE), &spa_stat.dvpp_va_size);
 	}
 
-	/* FIXME: UAF to be fix */
 	if (!is_local_group(spa->spg->id)) {
 		atomic_inc(&sp_overall_stat.spa_total_num);
 		atomic64_add(size, &sp_overall_stat.spa_total_size);
@@ -803,7 +802,6 @@ static void spa_dec_usage(struct sp_area *spa)
 		atomic64_sub(ALIGN(size, PMD_SIZE), &spa_stat.dvpp_va_size);
 	}
 
-	/* FIXME: UAF to be fix */
 	if (!is_local_group(spa->spg->id)) {
 		atomic_dec(&sp_overall_stat.spa_total_num);
 		atomic64_sub(spa->real_size, &sp_overall_stat.spa_total_size);
@@ -1581,6 +1579,16 @@ static void sp_group_insert_area(struct sp_group *spg, struct sp_area *spa)
 {
 	insert_sp_area(&spg->spa_root, spa, &spa->spg_link, spg_link);
 	atomic_inc(&spg->spa_num);
+	spa_inc_usage(spa);
+}
+
+/* the caller must hold spg->rw_lock */
+static void sp_group_delete_area(struct sp_group *spg, struct sp_area *spa)
+{
+	rb_erase(&spa->spg_link, &spg->spa_root);
+	RB_CLEAR_NODE(&spa->spg_link);
+	atomic_dec(&spa->spg->spa_num);
+	spa_dec_usage(spa);
 }
 
 /**
@@ -1736,7 +1744,6 @@ static struct sp_area *sp_area_alloc(unsigned long size, unsigned long flags,
 	spin_unlock(&mapping->sp_mapping_lock);
 	sp_group_insert_area(spg, spa);
 
-	spa_inc_usage(spa);
 	return spa;
 
 error:
@@ -1822,20 +1829,27 @@ static void sp_area_free(struct sp_area *spa)
 	if (spa->kva && !vmalloc_area_clr_flag(spa->kva, VM_SHAREPOOL))
 		pr_debug("clear spa->kva %ld is not valid\n", spa->kva);
 
-	spa_dec_usage(spa);
-
 	rb_erase(&spa->rb_node, &spm->area_root);
 	spin_unlock(&spm->sp_mapping_lock);
 	RB_CLEAR_NODE(&spa->rb_node);
 	kfree(spa);
 }
 
+static void sp_area_put_locked(struct sp_area *spa)
+{
+	if (atomic_dec_and_test(&spa->use_count)) {
+		if (!RB_EMPTY_NODE(&spa->spg_link))
+			sp_group_delete_area(spa->spg, spa);
+		sp_area_free(spa);
+	}
+}
+
 static void sp_area_put(struct sp_area *spa)
 {
 	if (atomic_dec_and_test(&spa->use_count)) {
 		down_write(&spa->spg->rw_lock);
-		rb_erase(&spa->spg_link, &spa->spg->spa_root);
-		atomic_dec(&spa->spg->spa_num);
+		if (!RB_EMPTY_NODE(&spa->spg_link))
+			sp_group_delete_area(spa->spg, spa);
 		up_write(&spa->spg->rw_lock);
 		sp_area_free(spa);
 	}
@@ -1843,6 +1857,8 @@ static void sp_area_put(struct sp_area *spa)
 
 void sp_area_drop(struct vm_area_struct *vma)
 {
+	struct sp_area *spa = vma->vm_private_data;
+
 	if (!(vma->vm_flags & VM_SHARE_POOL))
 		return;
 
@@ -1852,10 +1868,10 @@ void sp_area_drop(struct vm_area_struct *vma)
 	 * B calls sp_free() to free the same spa. So spa maybe NULL when A
 	 * calls this func later.
 	 */
-	if (!vma->vm_private_data)
+	if (!spa)
 		return;
 
-	sp_area_put(vma->vm_private_data);
+	sp_area_put(spa);
 }
 
 /*
@@ -1884,6 +1900,7 @@ static void sp_munmap(struct mm_struct *mm, unsigned long addr,
 	up_write(&mm->mmap_lock);
 }
 
+/* The caller should hold the write lock for spa->spg->rw_lock */
 static void __sp_free(struct sp_area *spa, struct mm_struct *stop)
 {
 	struct mm_struct *mm;
@@ -1980,9 +1997,7 @@ static int sp_free_inner(unsigned long addr, int spg_id, bool is_sp_free)
 		goto drop_spa;
 	}
 	spa->is_dead = true;
-	up_write(&spg->rw_lock);
 
-	down_read(&spg->rw_lock);
 	__sp_free(spa, NULL);
 	if (spa->type == SPA_TYPE_ALLOC) {
 		sp_fallocate(spa);
@@ -1990,7 +2005,7 @@ static int sp_free_inner(unsigned long addr, int spg_id, bool is_sp_free)
 		vmalloc_area_clr_flag(spa->kva, VM_SHAREPOOL);
 		spa->kva = 0;
 	}
-	up_read(&spg->rw_lock);
+	up_write(&spg->rw_lock);
 
 	if (current->mm != NULL)
 		sp_update_process_stat(current, false, spa);
@@ -2149,7 +2164,7 @@ static bool sp_alloc_fallback(struct sp_area *spa, struct sp_alloc_context *ac)
 	 * when we retry using normal pages.
 	 */
 	ac->have_mbind = false;
-	sp_area_put(spa);
+	sp_area_put_locked(spa);
 	return true;
 }
 
@@ -3305,7 +3320,6 @@ static void spa_stat_of_mapping_show(struct seq_file *seq, struct sp_mapping *sp
 	for (node = rb_first(&spm->area_root); node; node = rb_next(node)) {
 		spa = rb_entry(node, struct sp_area, rb_node);
 
-		/* FIXME: The spg here may have been freed */
 		if (spg_valid(spa->spg))  /* k2u to group */
 			seq_printf(seq, "%-10d ", spa->spg->id);
 		else  /* spg is dead */
-- 
2.17.1

